### PR TITLE
Move prompt templating to tokenizer

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
Prompt templating is often associated with the model or the finetuning task and needs to be carried over to inference/eval. To make this easier, we move prompt template as part of the tokenizer instead of the dataset.

Addresses #1343.

#### Changelog
- Update `ChatMLTemplate` for use in Qwen2Tokenizer. Simplify Qwen's tokenize_messages with `ChatMLTemplate`
- Add prompt template argument to all model tokenizers and builder functions
- Remove prompt template from all dataset classes and builders
- Add utility to define a prompt template from config via a string to the class name or a custom dict in the config
- Update all tests
- Add `__call__` to Qwen2Tokenizer as per #1353 

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](../tests/torchtune) for any new functionality
- [ ] update [docstrings](../docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

File: torchtune/data/_prompt_templates.py [hunk 1]
```diff
@@ -226,3 +242,29 @@ def __call__(
 
 Please see :class:`~torchtune.data.PromptTemplate` for full API arguments.
 """
+
+
+def _get_prompt_template(
```
Comment (joecummings) [inline torchtune/data/_prompt_templates.py:247]: This is super cool - we'll need some examples to make it clear to the user all the ways this can work.

File: torchtune/models/qwen2/_tokenizer.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
```
Comment (joecummings) [inline torchtune/models/qwen2/_tokenizer.py:10]: Is there a prompt template file missing or is that coming in a separate PR?

File: torchtune/data/_prompt_templates.py [hunk 3]
```diff
@@ -161,12 +163,24 @@ def __call__(
             The formatted list of messages
         """
         formatted_dialogue = []
-        for message in messages:
-            content = (
-                [{"type": "text", "content": self.template[message.role][0]}]
-                + message.content
-                + [{"type": "text", "content": self.template[message.role][1]}]
-            )
+        for index, message in enumerate(messages):
```
Comment (joecummings) [inline torchtune/data/_prompt_templates.py:166]: Was this a bug from before?

**Review by joecummings — APPROVED** [review]
I see no big issues with this - just one question and waiting for green CI.

File: torchtune/data/_prompt_templates.py (hunk 3 above)
Comment (RdoubleA) [reply torchtune/data/_prompt_templates.py:166]: kind of, see this discussion: https://github.com/pytorch/torchtune/pull/1019/files#r1621540801. it was also called out in the Qwen2 PR (#1143): 
>The official implemtation torchtune.data._chat_formats.ChatMLFormat will add an extra <|IM_END|> tag at the end of the last assistant message in generation (see [this discussion](https://github.com/pytorch/torchtune/pull/1019/files#r1621540801)), so we override the tokenize_message() method to fix this problem (will not add <|IM_END|> and EOS token if messages[-1].role == 'assistant' and not messages[-1].content)
